cunumeric/array.py

```diff
@@ -955,12 +955,101 @@ def byteswap(self, inplace=False):
             numpy_array = self.__array__(stacklevel=3).byteswap(inplace=False)
             return self.convert_to_cunumeric_ndarray(numpy_array, stacklevel=3)
 
-    @unimplemented
-    def choose(self, choices, out, mode="raise"):
-        numpy_array = self.__array__(stacklevel=3).choose(
-            choices=choices, out=out, mode=mode
-        )
-        return self.convert_to_cunumeric_ndarray(numpy_array, stacklevel=3)
+    def choose(self, choices, out=None, mode="raise", stacklevel=2):
+        a = self.convert_to_cunumeric_ndarray(self)
+        if out is not None:
+            out = out.convert_to_cunumeric_ndarray(out)
+        # choices can be passed as array or tuple of arrays.
+        # in case of the tuple, we will pass each array separately
+        is_tuple = isinstance(choices, tuple)
+        if is_tuple:
+            n = len(choices)
+            ch_dtype = choices[0].dtype
+        else:
+            n = choices.shape[0]
+            ch_dtype = choices.dtype
+
```

Comment:
It is not guaranteed that all choices arrays are the same type. We have to check that, and upcast them if necessary.

Reply:
What you are doing now:
```
ch_dtype = choices[0].dtype
choices = tuple(self.convert_to_cunumeric_ndarray(choices[i]).astype(ch_dtype), ...)
```
is almost correct, but casts everything to whatever the type of the first choice array is. NumPy will instead take the widest type from all the arguments:
```
>>> np.choose([0,1,0,1], (np.ones(4,dtype=int), np.zeros(4,dtype=float)))
array([1., 0., 1., 0.])
```
You should use the `find_common_type` method to find the widest type across the choice arrays and use that.